File: creche_roster_android/engine/src/main/kotlin/com/creche/roster/engine/Layout.kt
/**
 * Single source of truth for what a printed week looks like. Kotlin port of
 * creche_roster.layout. The app's roster screen and the PDF exporter both
 * render [weekGrid], so they can never disagree.
 */
package com.creche.roster.engine

import java.time.LocalDate

const val NCOLS = 2 + NDAYS + 2 // No, Name, Mon..Fri, break, empty
val COL_WIDTHS_CHARS: List<Int> = listOf(5, 18) + List(NDAYS) { 16 } + listOf(10, 10)

/** fill/color are hex without '#'. serif is used for the title only. */
data class CellStyle(
    val fill: String? = null,
    val color: String = "000000",
    val bold: Boolean = false,
    val italic: Boolean = false,
    val size: Int = 10,
    val align: String = "center",
    val border: Boolean = true,
    val serif: Boolean = false,
)

val STYLES: Map<String, CellStyle> = mapOf(
    "title" to CellStyle(color = "E04A32", bold = true, size = 22, align = "left", border = false, serif = true),
    "subtitle" to CellStyle(bold = true, size = 14, align = "left", border = false),
    "header" to CellStyle(fill = "D9D9D9", bold = true),
    "plain" to CellStyle(),
    "holiday" to CellStyle(fill = "C65B35", color = "F4CCB8"),
    "maternity" to CellStyle(fill = "2E9E57", color = "BFE6CB"),
    "off" to CellStyle(fill = "A9D18E"),
    "leave" to CellStyle(fill = "D9D9D9"),
    "blank" to CellStyle(border = false),
    "note" to CellStyle(color = "404040", italic = true, size = 9, align = "left", border = false),
    "note_warn" to CellStyle(fill = "FFEB9C", color = "7F6000", size = 9, align = "left", border = false),
    "note_bad" to CellStyle(fill = "FFC7CE", color = "9C0006", bold = true, size = 9, align = "left", border = false),
)

const val MAX_NOTES = 6

data class LCell(val text: String, val style: String)

data class GridRow(
    val cells: List<LCell>,
    // True: a single cell spanning every column.
    val merged: Boolean = false,
    // Relative height (1.0 = normal row).
    val height: Double = 1.0,
)

fun leaveStyle(text: String): String {
    val t = text.trim().lowercase()
    return when {
        t == "off" -> "off"
        t.contains("holiday") -> "holiday"
        t.contains("maternity") -> "maternity"
        else -> "leave"
    }
}

fun ordinal(n: Int): String {
    val suffix = if (n % 100 in 10..20) {
        "th"
    } else {
        when (n % 10) {
            1 -> "st"
            2 -> "nd"
            3 -> "rd"
            else -> "th"
        }
    }
    return "$n$suffix"
}

private val FULL_MONTHS = listOf(
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
)

private fun monthName(d: LocalDate) = FULL_MONTHS[d.monthValue - 1]

/** "21st – 25th September 2026" (or across months / years). */
fun dateRangeText(first: LocalDate, last: LocalDate): String {
    val dash = "–"
    return when {
        first.year == last.year && first.monthValue == last.monthValue ->
            "${ordinal(first.dayOfMonth)} $dash ${ordinal(last.dayOfMonth)} ${monthName(last)} ${last.year}"
        first.year == last.year ->
            "${ordinal(first.dayOfMonth)} ${monthName(first)} $dash ${ordinal(last.dayOfMonth)} ${monthName(last)} ${last.year}"
        else ->
            "${ordinal(first.dayOfMonth)} ${monthName(first)} ${first.year} $dash ${ordinal(last.dayOfMonth)} ${monthName(last)} ${last.year}"
    }
}

private val ORD_RE = Regex("""(\d+)(st|nd|rd|th)\b""")

/** Split "21st - 25th September" into runs; the boolean marks the raised suffix (st, nd, rd, th). */
fun ordinalRuns(text: String): List<Pair<String, Boolean>> {
    val runs = mutableListOf<Pair<String, Boolean>>()
    var pos = 0
    for (m in ORD_RE.findAll(text)) {
        val numEnd = m.groups[1]!!.range.last + 1
        runs.add(text.substring(pos, numEnd) to false)
        runs.add(m.groupValues[2] to true)
        pos = m.range.last + 1
    }
    runs.add(text.substring(pos) to false)
    return runs.filter { it.first.isNotEmpty() }
}

private fun merged(text: String, style: String, height: Double = 1.0): GridRow =
    GridRow(listOf(LCell(text, style)), merged = true, height = height)

fun noteHeight(text: String): Double {
    val lines = maxOf(1, (text.length + 89) / 90) // ceiling division
    return 0.75 * lines + 0.25
}

fun weekGrid(roster: Roster, wi: Int): List<GridRow> {
    val s = roster.inputs.settings
    val week = roster.weeks[wi]
    val rows = mutableListOf<GridRow>()

    rows.add(merged(s.title, "title", 2.0))
    rows.add(merged(dateRangeText(week.days.first(), week.days.last()), "subtitle", 1.6))
    rows.add(merged("", "blank", 0.7))

    if (s.dayHeaders) {
        val headerCells = mutableListOf(LCell("", "header"), LCell("Name", "header"))
        week.days.forEachIndexed { i, d ->
            headerCells.add(LCell("${DAY_NAMES[i]} ${d.dayOfMonth} ${MONTH_ABBR_LOCAL[d.monthValue - 1]}", "header"))
        }
        headerCells.add(LCell("Break", "header"))
        headerCells.add(LCell("", "header"))
        rows.add(GridRow(headerCells, height = 1.25))
    }

    var number = 0
    for ((key, st) in roster.staffKeys) {
        if (st.role == "blank") { // spacer row: bordered, empty, unnumbered
            rows.add(GridRow(List(NCOLS) { LCell("", "plain") }, height = 1.25))
            continue
        }
        val label: String
        if (st.number.trim() == "-") {
            label = ""
        } else {
            val typed = st.number.trim().toIntOrNull()
            number = typed ?: (number + 1)
            label = number.toString()
        }
        val dayCells = week.days.map { d ->
            val a = week.cells[key to d] ?: Assignment("blank")
            LCell(a.text, if (a.kind == "leave") leaveStyle(a.text) else "plain")
        }
        val hasContent = st.name.isNotEmpty() || dayCells.any { it.text.isNotEmpty() }
        val cells = mutableListOf(LCell(label, "plain"), LCell(st.name, "plain"))
        cells.addAll(dayCells)
        cells.add(LCell(if (hasContent) s.breakText else "", "plain"))
        cells.add(LCell("", "plain"))
        rows.add(GridRow(cells, height = 1.25))
    }

    // Anything that breaks a hard rule is printed under the table so nobody
    // hands out a broken roster without knowing.
    val bad = roster.checks.filter { it.week == wi && it.level == "BREACH" }
    for (c in bad.take(MAX_NOTES)) {
        val text = "BREACH: ${c.message}"
        rows.add(merged(text, "note_bad", noteHeight(text)))
    }
    if (bad.size > MAX_NOTES) {
        rows.add(merged("+ ${bad.size - MAX_NOTES} more (see the Checks screen)", "note_bad"))
    }
    return rows
}

private val MONTH_ABBR_LOCAL = listOf(
    "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
)

File: creche_roster_app/webapp/functions/creche_roster/layout.py
"""Single source of truth for what a printed week looks like.

This reproduces the old paper roster: a red "STAFF ROSTER" title, the date
range, then one table with the columns

    No | Name | Mon | Tue | Wed | Thu | Fri | break | (empty)

Leave is coloured: Holiday orange-red, Maternity Leave green, OFF light
green. A shift is coloured amber when that day it was moved to cover for a
colleague's leave (the "Cover adjusted" / "Closing fallback" checks) -
otherwise shifts carry no colour. Times are shown 12-hour style (7:30 - 4:30).

Both the Excel writer and the PDF writer render `week_grid(...)`, so the two
files always show exactly the same names, times, colours and notes.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Dict, List, Tuple

from .models import DAY_NAMES, NDAYS, Roster

NCOLS = 2 + NDAYS + 2  # No, Name, Mon..Fri, break, empty
COL_WIDTHS_CHARS = [5, 18] + [16] * NDAYS + [10, 10]  # Excel column widths

# fill/color are hex without '#'. `serif` is used for the title only.
_BASE = dict(fill=None, color="000000", bold=False, italic=False, size=10, align="center", border=True, serif=False)


def _style(**kw) -> dict:
    d = dict(_BASE)
    d.update(kw)
    return d


STYLES: Dict[str, dict] = {
    "title": _style(color="E04A32", bold=True, size=22, align="left", border=False, serif=True),
    "subtitle": _style(bold=True, size=14, align="left", border=False),
    "header": _style(fill="D9D9D9", bold=True),
    "plain": _style(),
    "holiday": _style(fill="C65B35", color="F4CCB8"),
    "maternity": _style(fill="2E9E57", color="BFE6CB"),
    "off": _style(fill="A9D18E"),
    "leave": _style(fill="D9D9D9"),
    "adjusted": _style(fill="FCE4A0", color="7F5F00", bold=True),
    "blank": _style(border=False),
    "note": _style(color="404040", italic=True, size=9, align="left", border=False),
    "note_warn": _style(fill="FFEB9C", color="7F6000", size=9, align="left", border=False),
    "note_bad": _style(fill="FFC7CE", color="9C0006", bold=True, size=9, align="left", border=False),
}

MAX_NOTES = 6


@dataclass
class LCell:
    text: str
    style: str


@dataclass
class Row:
    cells: List[LCell]
    merged: bool = False  # True: a single cell spanning every column
    height: float = 1.0  # relative height (1.0 = normal row)


def leave_style(text: str) -> str:
    t = text.strip().lower()
    if t == "off":
        return "off"
    if "holiday" in t:
        return "holiday"
    if "maternity" in t:
        return "maternity"
    return "leave"


def ordinal(n: int) -> str:
    if 10 <= n % 100 <= 20:
        suffix = "th"
    else:
        suffix = {1: "st", 2: "nd", 3: "rd"}.get(n % 10, "th")
    return f"{n}{suffix}"


def date_range_text(first, last) -> str:
    """'21st – 25th September 2026' (or across months / years)."""
    dash = "–"
    if (first.year, first.month) == (last.year, last.month):
        return f"{ordinal(first.day)} {dash} {ordinal(last.day)} {last:%B %Y}"
    if first.year == last.year:
        return f"{ordinal(first.day)} {first:%B} {dash} {ordinal(last.day)} {last:%B %Y}"
    return f"{ordinal(first.day)} {first:%B %Y} {dash} {ordinal(last.day)} {last:%B %Y}"


_ORD_RE = re.compile(r"(\d+)(st|nd|rd|th)\b")


def ordinal_runs(text: str) -> List[Tuple[str, bool]]:
    """Split '21st - 25th September' into runs; True marks the raised suffix (st, nd, rd, th)."""
    runs: List[Tuple[str, bool]] = []
    pos = 0
    for m in _ORD_RE.finditer(text):
        runs.append((text[pos:m.end(1)], False))
        runs.append((m.group(2), True))
        pos = m.end()
    runs.append((text[pos:], False))
    return [r for r in runs if r[0]]


def _merged(text: str, style: str, height: float = 1.0) -> Row:
    return Row([LCell(text, style)], merged=True, height=height)


def _note_height(text: str) -> float:
    lines = max(1, -(-len(text) // 90))  # ceiling division
    return 0.75 * lines + 0.25


def week_grid(roster: Roster, wi: int) -> List[Row]:
    s = roster.inputs.settings
    week = roster.weeks[wi]
    rows: List[Row] = []

    rows.append(_merged(s.title, "title", 2.0))
    rows.append(_merged(date_range_text(week.days[0], week.days[-1]), "subtitle", 1.6))
    rows.append(_merged("", "blank", 0.7))

    if s.day_headers:
        rows.append(
            Row(
                [LCell("", "header"), LCell("Name", "header")]
                + [LCell(f"{DAY_NAMES[i]} {d.day} {d:%b}", "header") for i, d in enumerate(week.days)]
                + [LCell("Break", "header"), LCell("", "header")],
                height=1.25,
            )
        )

    number = 0
    for key, st in roster.staff_keys:
        if st.role == "blank":  # spacer row: bordered, empty, unnumbered
            rows.append(Row([LCell("", "plain") for _ in range(NCOLS)], height=1.25))
            continue
        if st.number.strip() == "-":
            label = ""
        else:
            number = int(st.number) if st.number.strip().isdigit() else number + 1
            label = str(number)
        day_cells: List[LCell] = []
        for d in week.days:
            a = week.cells[(key, d)]
            if a.kind == "leave":
                style = leave_style(a.text)
            elif a.kind == "shift" and a.adjusted:
                style = "adjusted"
            else:
                style = "plain"
            day_cells.append(LCell(a.text, style))
        has_content = bool(st.name) or any(c.text for c in day_cells)
        cells = (
            [LCell(label, "plain"), LCell(st.name, "plain")]
            + day_cells
            + [LCell(s.break_text if has_content else "", "plain"), LCell("", "plain")]
        )
        rows.append(Row(cells, height=1.25))

    # Anything that breaks a hard rule is printed under the table so nobody
    # hands out a broken roster without knowing.
    bad = [c for c in roster.checks if c.week == wi and c.level == "BREACH"]
    for c in bad[:MAX_NOTES]:
        text = "BREACH: " + c.message
        rows.append(_merged(text, "note_bad", _note_height(text)))
    if len(bad) > MAX_NOTES:
        rows.append(_merged(f"+ {len(bad) - MAX_NOTES} more (see the Checks tab)", "note_bad"))
    return rows
